packages/components/src/components/Dropdown/Menu.tsx

```diff
@@ -0,0 +1,412 @@
+            const getPrevIndex = (current: number) =>
+                current > 0 ? current - 1 : groupItemIds.length - 1;
+            nextIndex = getPrevIndex(nextIndex);
+
+            // skip disabled items
+            while (groupItemIds[nextIndex].isDisabled) {
+                nextIndex = getPrevIndex(nextIndex);
+            }
+        } else if (keyboardKey === 'ArrowDown') {
+            const getNextIndex = (current: number) =>
+                current < groupItemIds.length - 1 ? current + 1 : 0;
+            nextIndex = getNextIndex(nextIndex);
+
+            // skip disabled items
+            while (groupItemIds[nextIndex].isDisabled) {
+                nextIndex = getNextIndex(nextIndex);
+            }
+        }
+
+        return nextIndex;
+    };
+
+const getDefaultFocusItemIndex = (
+    isKeyboardControlled: MenuProps['isKeyboardControlled'],
+    items: MenuProps['items'],
+    addon: MenuProps['addon'],
+) => {
+    if (isKeyboardControlled && items?.length) {
+        return addon ? 1 : 0;
+    }
+    return null;
+};
+
+export type MenuAlignment = 'left' | 'right' | 'top-left' | 'top-right';
+
+export interface MenuProps {
+    items?: GroupedMenuItems[];
+    content?: React.ReactNode;
+    alignMenu?: MenuAlignment;
+    coords?: Coords;
+    addon?: Omit<AddonProps, 'setToggled'>;
+    isKeyboardControlled?: boolean;
+    setToggled: (toggled: boolean) => void;
+}
+
+export const Menu = forwardRef<HTMLUListElement, MenuProps>(
+    ({ items, content, setToggled, alignMenu, coords, addon, isKeyboardControlled }, ref) => {
+        const [focusedItemIndex, setFocusedItemIndex] = useState(
+            getDefaultFocusItemIndex(isKeyboardControlled, items, addon),
+        );
+
+        const visibleItems = items?.map(group => ({
+            ...group,
+            options: group.options.filter(item => !item.isHidden),
+        }));
+
+        const groupItemIds = visibleItems?.reduce((ids, group, groupIndex) => {
+            const groupIds = group.options.map(({ isDisabled }, index) => ({
+                id: `${groupIndex}.${index}`,
+                isDisabled,
+            }));
+
+            return [...ids, ...groupIds];
+        }, [] as Array<{ id: string; isDisabled?: boolean }>);
+
+        if (addon) {
+            groupItemIds?.unshift({ id: 'addon' });
+        }
+
+        useEffect(() => {
+            const handleKeyDown = (e: KeyboardEvent) => {
+                if (groupItemIds && groupItemIds.length > 0 && focusedItemIndex !== null) {
+                    setFocusedItemIndex(getNextIndex(e.key, groupItemIds));
+                }
+            };
+
+            if (focusedItemIndex !== null && groupItemIds?.length) {
+                document.addEventListener('keydown', handleKeyDown);
+
+                return () => {
+                    document.removeEventListener('keydown', handleKeyDown);
+                };
+            }
+        }, []); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
I was trying to deal with the dependency issue and came up with this. Of course it can be polished further.

```ts
const getNewIndex =
    (keyboardKey: string, prevIndex: number, nextIndex: number) =>
    (currentIndex: number | null) => {
        if (currentIndex === null) {
            return null;
        }
        if (keyboardKey === 'ArrowUp') {
            return prevIndex;
        }
        if (keyboardKey === 'ArrowDown') {
            return nextIndex;
        }
        return currentIndex;
    };
```

```ts
        const lastIndex = groupItemIds ? groupItemIds.length - 1 : 0;

        const getNextIndex = () => {
            if (groupItemIds && focusedItemIndex !== null && focusedItemIndex < lastIndex)
                for (let i = focusedItemIndex + 1; i <= lastIndex; i++) {
                    if (!groupItemIds[i].isDisabled) {
                        return i;
                    }
                }
            return 0;
        };
        const getPreviousIndex = () => {
            if (groupItemIds && focusedItemIndex !== null)
                for (let i = focusedItemIndex - 1; i >= 0; i--) {
                    if (!groupItemIds[i].isDisabled) {
                        return i;
                    }
                }
            return lastIndex;
        };

        const previousIndex = getPreviousIndex();
        const nextIndex = getNextIndex();

        useEffect(() => {
            const handleKeyDown = (e: KeyboardEvent) => {
                if (focusedItemIndex !== null) {
                    setFocusedItemIndex(getNewIndex(e.key, previousIndex, nextIndex));
                }
            };

            if (focusedItemIndex !== null) {
                document.addEventListener('keydown', handleKeyDown);

                return () => {
                    document.removeEventListener('keydown', handleKeyDown);
                };
            }
        }, [focusedItemIndex, nextIndex, previousIndex]);
```

Reply:
what problem does it solve exactly? Actually, I have put the dependencies back there and it all works fine now 😓

Reply:
Also what I think that in your code if an item at `0` or `lastIndex` is  disabled, it's not skipped :)

Reply:
😭 I thought you didn't include the deps to prevent some loop due to the array being used as a dependency.